### PR TITLE
[wip] add DFR chapter

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Figured it'd be nice to highlight how to design reusable components; further zooming in on @maxogden's comments in IRC

This provides the potential to zoom into the design decisions underlying `choo`; the full interop it has with any methods that operates on raw DOM and highlighting patterns to prevent completely getting tangled up in another framework, yet again.

Also @shama got me a bit worried about completely buying into frameworks haha (as per https://github.com/yoshuawuyts/choo/issues/152) - feel like having a guide might align mindsets or smth like that.

Anyway this is very much wip - but I'm off to bed so sharing what I got so far. Night!

### CODE REVIEW

File: designing-for-reusability.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+This guide is intended to show some patterns you can apply to your code in
+order to make it more reusable, and outlast the cycle of frameworks.
+
+Note that in this guide the words "module" and "package" will be used
+interchangeably.
+
+## Directory structure
+In order to facilitate reusability, one must plan for it. Often the hardest
+part about modularity is becoming comfortable with determining where the
+boundries are. As a rule of thumb: any piece of logic that doesn't directly
```
Comment (shama) [inline designing-for-reusability.md]: `/boundries/boundaries/`

File: designing-for-reusability.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+Say we have a form element. We want it to trigger two types of validation: when
+input occurs we want to validate it on the client. When submit is clicked we
+want it to submit the data and validate it on the server. To create a
+generalized component we'd do:
+```js
+// ./pages/main.js
+const html = require('choo/html')
+const Form = require('../components/form')
+
+const renderForm = Form(formInput, {
```
Comment (timwis) [inline designing-for-reusability.md]: I think you mean to use a function signature here, ie `const renderForm = (formInput) => Form(formInput, {`

File: designing-for-reusability.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    <main>
+      ${renderForm(formInput)}
+    </main>
+  `
+}
+```
+
+```js
+// ./components/form.js
+const getFormData = require('form-data')
```
Comment (timwis) [inline designing-for-reusability.md]: I think you mean `require('get-form-data')` (for [this package](https://www.npmjs.com/package/get-form-data))

File: designing-for-reusability.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+const html = require('choo/html')
+const assert = require('assert')
+
+module.exports = function formComponent (evMap) {
+  assert.equal(typeof evMap, 'object')
```
Comment (timwis) [inline designing-for-reusability.md]: Perhaps for the same reason we're not too ES6-heavy in handbooks, the not-very-common/familiar concept of inlining `assert`s should be introduced elsewhere?

File: designing-for-reusability.md (hunk 3 above)
Comment (yoshuawuyts) [inline designing-for-reusability.md]: haha, yessss! - wrote this offline so oops :D

File: designing-for-reusability.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+  return function (input) {
```
Comment (timwis) [inline designing-for-reusability.md]: Er, am I reading this wrong? Why is it returning a function?

File: designing-for-reusability.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  function onInput (e) {
+    const data = e.value
+    evMap.onInput(data)
+  }
+
+  function onSubmit (e) {
+    e.preventDefault()
+    const data = getFormData(e)
```
Comment (timwis) [inline designing-for-reusability.md]: I believe you mean to use `e.target` here

File: designing-for-reusability.md (hunk 5 above)
Comment (yoshuawuyts) [inline designing-for-reusability.md]: It's being thunked; to have it like have an "init" stage and a "render" stage; the API for the consumers looks kinda nice this way, and there's no way of accidentally redeclaring the objects inline which would make it super slow. Does that make sense? Should we explain that better?

**Comment by yoshuawuyts** [comment]
Yeah cool, addressed all notes - per @timwis's suggestion on IRC let's merge and edit from there - neat to like have this available already :sparkles:

**Pull request merged**